### PR TITLE
add extension fallback

### PR DESCRIPTION
**Pull request by Leouarz**
# Pull Request type

Please add the labels corresponding to the type of changes your PR introduces:

- [x] Refactor

## Description

It adds a last resort mechanism in case, for some reason, the commitment or grid construction fails.
Companion PR in avail-core : https://github.com/availproject/avail-core/pull/74

## Related Issues

[Closes engineering#435](https://github.com/availproject/engineering/issues/435)

## Checklist
- [x] I have performed a self-review of my own code.
- [x] The tests pass successfully with `cargo test`.
- [x] The code was formatted with `cargo fmt`.
- [x] The code compiles with no new warnings with `cargo build --release` and `cargo build --release --features runtime-benchmarks`.
- [x] The code has no new warnings when using `cargo clippy`.
- [x] If this change affects documented features or needs new documentation, I have created a PR with a [documentation update](https://github.com/availproject/availproject.github.io).

### CODE REVIEW

File: pallets/system/src/header_builder.rs [hunk 1]
```diff
@@ -104,51 +108,129 @@ fn corrupt_commitment(block_number: u32, commitment: &mut Vec<u8>) {
...
+
+	// Build the grid
+	let timer = std::time::Instant::now();
+	let maybe_grid = build_grid(app_extrinsics, block_length, seed);
+	// Evaluation Grid Build Time Metrics
+	HeaderExtensionBuilderMetrics::observe_evaluation_grid_build_time(timer.elapsed());
+	let grid = match maybe_grid {
+		Ok(res) => res,
+		Err(message) => {
+			log::error!("NODE_CRITICAL_ERROR_001 - A critical error has occured: {message:?}.");
```
Comment (prabal-banerjee) [inline pallets/system/src/header_builder.rs:209]: Can we have some other error code? Looks verbose. Something like this (although not equivalent)?
https://github.com/availproject/avail/wiki/Transaction-Custom-IDs

Comment (prabal-banerjee) [reply pallets/system/src/header_builder.rs:209]: And then a corresponding error on the wiki,

**Review by prabal-banerjee — APPROVED** [review]

File: pallets/system/src/header_builder.rs (hunk 1 above)
Comment (Leouarz) [reply pallets/system/src/header_builder.rs:209]: We chose to not introduce the same type of error for multiple reasons : 
- This is a console only error
- This is not a transaction error that need to be returned to the user so we cannot use the same feat
- It needs to be easily findable

**Review by markopoloparadox — APPROVED** [review]

**Review by markopoloparadox — APPROVED** [review]

**Review by ToufeeqP — APPROVED** [review]

**Review by sh3ll3x3c — APPROVED** [review]

**Review by Aphoh — CHANGES_REQUESTED** [review]
Looks pretty good! Just one more place it would be good to do an error check instead of a panic. It would be really nice to put something like [this](https://github.com/dtolnay/no-panic) in there too, but my guess is that my code downsteam doesn't pass this kind of scrutiny 😅

File: pallets/system/src/header_builder.rs [hunk 2]
```diff
@@ -104,51 +108,129 @@ fn corrupt_commitment(block_number: u32, commitment: &mut Vec<u8>) {
...
+		.map_err(|e| format!("Make polynomial grid failed: {e:?}"))?;
+
+	let extended_grid = poly_grid
 		.extended_commitments(&*PMP, 2)
-		.expect("Extended commitments cannot fail")
+		.map_err(|e| format!("Grid extension failed: {e:?}"))?;
+
+	let commitment = extended_grid
 		.iter()
 		.flat_map(|c| c.to_bytes().expect("Commitment serialization cannot fail"))
```
[the remark this answers is not in the document]
Comment (Leouarz) [reply pallets/system/src/header_builder.rs:148]: It can be done, but this literally should never fail, wdy think ?

**Pull request merged**